diffrax/_solver/quicsort.py

```diff
@@ -0,0 +1,250 @@
+        b = jnp.stack([c4 / 720, -c3 / 120, c2 / 24, -c / 6, one / 2, zero], axis=-1)
+
+        with jax.numpy_rank_promotion("allow"):
+            beta_lr1 = lr1_pows * jnp.expand_dims(beta, axis=c.ndim)
+            a_lr1 = lr1_pows * jnp.expand_dims(a, axis=c.ndim)
+            # b needs an extra power of lr1 (just work out the expansion to see why)
+            b_lr1 = lr1_pows * lr1 * jnp.expand_dims(b, axis=c.ndim)
+        assert beta_lr1.shape == a_lr1.shape == b_lr1.shape == c.shape + (3, 6)
+
+        # a_third = (1 - exp(-1/3 * gamma * h))/gamma
+        a_third = jnp.stack(
+            [c4 / 29160, -c3 / 1944, c2 / 162, -c / 18, one / 3, zero], axis=-1
+        )
+        a_third = jnp.expand_dims(a_third, axis=c.ndim)
+        a_div_h = jnp.stack(
+            [-c5 / 720, c4 / 120, -c3 / 24, c2 / 6, -c / 2, one], axis=-1
+        )
+        a_div_h = jnp.expand_dims(a_div_h, axis=c.ndim)
+        assert a_third.shape == a_div_h.shape == c.shape + (1, 6)
+
+        return _QUICSORTCoeffs(
+            beta_lr1=beta_lr1,
+            a_lr1=a_lr1,
+            b_lr1=b_lr1,
+            a_third=a_third,
+            a_div_h=a_div_h,
+        )
+
+    def _compute_step(
+        self,
+        h: RealScalarLike,
+        levy: AbstractSpaceTimeTimeLevyArea,
+        x0: UnderdampedLangevinX,
+        v0: UnderdampedLangevinX,
+        uld_args: ULDArgs,
+        coeffs: _QUICSORTCoeffs,
+        rho: UnderdampedLangevinX,
+        prev_f: UnderdampedLangevinX,
+    ) -> tuple[UnderdampedLangevinX, UnderdampedLangevinX, UnderdampedLangevinX, None]:
+        dtypes = jtu.tree_map(jnp.result_type, x0)
+        w: UnderdampedLangevinX = jtu.tree_map(jnp.asarray, levy.W, dtypes)
+        hh: UnderdampedLangevinX = jtu.tree_map(jnp.asarray, levy.H, dtypes)
+        kk: UnderdampedLangevinX = jtu.tree_map(jnp.asarray, levy.K, dtypes)
+
+        gamma, u, f = uld_args
+
+        def _extract_coeffs(coeff, index):
+            return jtu.tree_map(lambda arr: arr[..., index], coeff)
+
+        beta_l = _extract_coeffs(coeffs.beta_lr1, 0)
+        beta_r = _extract_coeffs(coeffs.beta_lr1, 1)
+        beta_1 = _extract_coeffs(coeffs.beta_lr1, 2)
+        a_l = _extract_coeffs(coeffs.a_lr1, 0)
+        a_r = _extract_coeffs(coeffs.a_lr1, 1)
+        a_1 = _extract_coeffs(coeffs.a_lr1, 2)
+        b_l = _extract_coeffs(coeffs.b_lr1, 0)
+        b_r = _extract_coeffs(coeffs.b_lr1, 1)
+        b_1 = _extract_coeffs(coeffs.b_lr1, 2)
+        a_third = _extract_coeffs(coeffs.a_third, 0)
+        a_div_h = _extract_coeffs(coeffs.a_div_h, 0)
+
+        rho_w_k = (rho**ω * (w**ω - 12 * kk**ω)).ω
+        uh = (u**ω * h).ω
+        v_tilde = (v0**ω + rho**ω * (hh**ω + 6 * kk**ω)).ω
+
+        x1 = (x0**ω + a_l**ω * v_tilde**ω + b_l**ω * rho_w_k**ω).ω
+        f1uh = (f(x1) ** ω * uh**ω).ω
+
+        x2 = (
+            x0**ω + a_r**ω * v_tilde**ω + b_r**ω * rho_w_k**ω - a_third**ω * f1uh**ω
+        ).ω
+        f2uh = (f(x2) ** ω * uh**ω).ω
```

Comment:
Let's not do this now -- leave it for a follow-up PR if you have the energy (+we can then more easily verify correctness) -- but this would be better done as a little length-2 scan, so that we only need to compile `f` once.

See how e.g. `_step_size_controller/adaptive.py::_select_initial_step` uses `eqxi.scan_trick`.

Reply:
Done! It's a very clever trick indeed 😊.

Just how significant are the speedups from this? I imagine a more complicated piece of code (so `eqxi.scan_trick` in this case) brings a computational overhead of its own, so this is surely a tradeoff. Sometimes it's really not obvious to me if it is worth having significantly more convoluted code to achieve some speedup. How do you decide? Do you benchmark these things?

Reply:
The decreased readability is indeed substantial. It's unfortunate.

Generally this is worth doing for anything involving the user-provided vector. These sometimes get very large indeed; every extra trace can increase compile time by several minutes.

Reply:
Oh, so the point is that it's user-supplied, so the worst case scenario can be arbitrarily expensive, whereas we know that the overhead of the code is not that big. Makes sense, thanks!